# Pandas Revision — The `mpg` Dataset

398 cars sold in the United States between 1970 and 1982, with fuel economy, engine specs,
weight and country of origin. It is small enough to check by eye and messy enough to be real.

Eighteen questions. They run roughly in the order we covered the topics, so if you are stuck
on Q11 the answer is not in Q17 — it is in your notes from the `idxmax` section.

Two of these questions (**Q10** and **Q17**) are not about syntax. You will get a number in
about thirty seconds, and then the actual work begins: deciding whether that number means
what it appears to mean. Those are the ones worth your time.

**The columns**

| Column | Meaning |
|--------|---------|
| `mpg` | Miles per gallon — higher is more efficient |
| `cylinders` | Engine cylinders (3, 4, 5, 6 or 8) |
| `displacement` | Engine size, cubic inches |
| `horsepower` | Engine power |
| `weight` | Kerb weight, pounds |
| `acceleration` | Seconds from 0 to 60 mph — *lower* is faster |
| `model_year` | Two-digit year: 70 = 1970 |
| `origin` | `usa`, `japan` or `europe` |
| `name` | Make and model |

In [1]:
import pandas as pd

cars = pd.read_csv('mpg.csv')
cars.head()

,mpg,cylinders,displacement,horsepower,weight,acceleration,model_year,origin,name
0,18.0,8,307.0,130.0,3504,12.0,70,usa,chevrolet chevelle malibu
1,15.0,8,350.0,165.0,3693,11.5,70,usa,buick skylark 320
2,18.0,8,318.0,150.0,3436,11.0,70,usa,plymouth satellite
3,16.0,8,304.0,150.0,3433,12.0,70,usa,amc rebel sst
4,17.0,8,302.0,140.0,3449,10.5,70,usa,ford torino


---
## Q1. Inspect the data

Before touching anything, find out what you are holding.

1. How many rows and columns?
2. What is the dtype of each column?
3. Which columns have missing values, and how many?

In [7]:
# your code here
print(cars.shape)
print(cars.dtypes)
print(cars.isnull().sum())

(398, 9)
mpg             float64
cylinders         int64
displacement    float64
horsepower      float64
weight            int64
acceleration    float64
model_year        int64
origin           object
name             object
dtype: object
mpg             0
cylinders       0
displacement    0
horsepower      6
weight          0
acceleration    0
model_year      0
origin          0
name            0
dtype: int64


---
## Q2. Select columns

Show only `name`, `mpg` and `origin`, for the first 5 rows.

In [8]:
# your code here
cars[['name', 'mpg', 'origin']].head()

,name,mpg,origin
0,chevrolet chevelle malibu,18.0,usa
1,buick skylark 320,15.0,usa
2,plymouth satellite,18.0,usa
3,amc rebel sst,16.0,usa
4,ford torino,17.0,usa


---
## Q3. Filter rows

How many cars manage more than 30 mpg? Of those, how many are Japanese?

In [14]:
# your code here
print(cars[(cars['mpg']>30) & (cars['origin']=='japan')])

      mpg  cylinders  displacement  horsepower  weight  acceleration  \
53   31.0          4          71.0        65.0    1773          19.0   
54   35.0          4          72.0        69.0    1613          18.0   
129  31.0          4          79.0        67.0    1950          19.0   
131  32.0          4          71.0        65.0    1836          21.0   
144  31.0          4          76.0        52.0    1649          16.5   
145  32.0          4          83.0        61.0    2003          19.0   
181  33.0          4          91.0        53.0    1795          17.5   
198  33.0          4          91.0        53.0    1795          17.4   
204  32.0          4          85.0        70.0    1990          17.0   
216  31.5          4          98.0        68.0    2045          18.5   
220  33.5          4          85.0        70.0    1945          16.8   
246  32.8          4          78.0        52.0    1985          19.4   
247  39.4          4          85.0        70.0    2070          

---
## Q4. Handle missing values

`horsepower` has blanks.

1. Which origins do the missing rows belong to?
2. Fill the blanks with the median horsepower.
3. Confirm there are no blanks left.

In [27]:
# your code here
cars[cars['horsepower'].isnull()]['origin']
cars['horsepower'] = cars['horsepower'].fillna(cars['horsepower'].median())
cars['horsepower'].isnull().sum()

np.int64(0)

---
## Q5. Create a new column

Power-to-weight ratio is a better predictor of how a car *feels* than horsepower alone.

Add a column `power_to_weight` = horsepower ÷ weight. Then show the 5 cars with the highest ratio.

In [32]:
# your code here
cars['power_to_weight'] = cars['horsepower']/cars['weight']
print(cars.sort_values('power_to_weight', ascending=False).head())

      mpg  cylinders  displacement  horsepower  weight  acceleration  \
13   14.0          8         455.0       225.0    3086          10.0   
116  16.0          8         400.0       230.0    4278           9.5   
330  40.9          4          85.0        93.5    1835          17.3   
8    14.0          8         455.0       225.0    4425          10.0   
23   26.0          4         121.0       113.0    2234          12.5   

     model_year  origin                     name  horspower  power_to_weight  
13           70     usa  buick estate wagon (sw)      225.0         0.072910  
116          73     usa       pontiac grand prix      230.0         0.053763  
330          80  europe     renault lecar deluxe       93.5         0.050954  
8            70     usa         pontiac catalina      225.0         0.050847  
23           70  europe                 bmw 2002      113.0         0.050582  


---
## Q6. Count with value_counts

1. How many cars come from each origin?
2. How many cars have each cylinder count?
3. Show the origin breakdown as percentages instead of counts.

In [45]:
# your code here
print(cars['origin'].value_counts())
print(cars['cylinders'].value_counts())
(cars['origin'].value_counts(normalize=True)*100).round(2)

origin
usa       249
japan      79
europe     70
Name: count, dtype: int64
cylinders
4    204
8    103
6     84
3      4
5      3
Name: count, dtype: int64


origin
usa       62.56
japan     19.85
europe    17.59
Name: proportion, dtype: float64

---
## Q7. Named aggregation

Build one table with a row per origin and these columns:
`cars`, `avg_mpg`, `avg_weight`, `avg_cylinders`. Round to 1 decimal.

In [49]:
# your code here
summary = cars.groupby('origin').agg(
    cars = ('name','count'),
    avg_mpg = ('mpg', 'mean'),
    avg_weight = ('weight', 'mean'),
    avg_cylinders = ('cylinders', 'mean')
)
summary

,cars,avg_mpg,avg_weight,avg_cylinders
origin,,,,
europe,70,27.891429,2423.300000,4.157143
japan,79,30.450633,2221.227848,4.101266
usa,249,20.083534,3361.931727,6.248996


---
## Q8. size vs count

For each origin, show how many cars there are **and** how many have a horsepower recorded.

*Run this on a fresh copy of the data — if you filled the blanks in Q4, they are gone.*

In [60]:
# your code here
nc = cars.copy()
su = nc.groupby('origin').agg(
    total_cars = ('origin', 'size'),
    horsepower_recorded = ('horsepower','count')
)
su

,total_cars,horsepower_recorded
origin,,
europe,70,70
japan,79,79
usa,249,249


---
## Q9. Group by two columns

Average mpg for every combination of `origin` and `cylinders`. Then reshape it with
`.unstack()` so origins are rows and cylinder counts are columns.

In [61]:
# your code here
cars.groupby(['origin', 'cylinders'])['mpg'].mean().unstack()

cylinders,3,4,5,6,8
origin,,,,,
europe,NaN,28.411111,27.366667,20.100000,NaN
japan,20.55,31.595652,NaN,23.883333,NaN
usa,NaN,27.840278,NaN,19.663514,14.963107


---
## Q10. ⭐ The confound

Someone shows you this:

> *Japanese cars average 30.5 mpg. American cars average 20.1. Japanese engineering is
> simply better at fuel efficiency.*

Both numbers are correct — you computed them yourself in Q7.

**Is the conclusion correct?** Investigate. Compare like with like, and state clearly what
the data does and does not support.

In [ ]:
# your code here


---
## Q11. idxmax

Which specific car has the best fuel economy in each origin? Show the car's name and mpg,
not just the number.

In [ ]:
# your code here


---
## Q12. astype and to_csv

1. Convert `origin` to the `category` dtype and report the memory saved.
2. Save your Q7 summary table to `origin_summary.csv` — with the origin column intact.
3. Read the file back and confirm nothing was lost.

In [ ]:
# your code here


---
## Q13. Sorting

Show the 10 heaviest cars, with name, weight and mpg. What do you notice about their mpg?

In [ ]:
# your code here


---
## Q14. Compound filter

Find American cars that are 4-cylinder, from model year 1978 or later, and manage more than
25 mpg. How many are there? Show the 5 most efficient.

In [ ]:
# your code here


---
## Q15. String methods

The `name` column holds make and model together, like `chevrolet chevelle malibu`.

1. Extract the brand — the first word — into a new `brand` column.
2. List every distinct brand.
3. Something is wrong with that list. Find the problem and fix it.

In [ ]:
# your code here


In [ ]:
# your code here


---
## Q16. Brand summary

Using your cleaned `brand` column: for every brand with at least 15 cars, show the number of
cars and the average mpg, sorted from most to least efficient.

In [ ]:
# your code here


---
## Q17. ⭐ Change over time

1. What was the average mpg of the 1970 fleet? The 1982 fleet?
2. Someone concludes: *"Japanese cars got much more efficient over the decade — that is why
   they took the market."* Check it. Did Japanese cars improve more than everyone else's?

In [ ]:
# your code here


In [ ]:
# your code here


---
## Q18. idxmin

For each cylinder count, which car has the *worst* fuel economy?

In [ ]:
# your code here


---
## When you are done

Q10 and Q17 are the two that matter. Anyone can look up `.agg()` later; recognising that an
average is comparing two different populations is the skill that does not come back on its own.

If you finished early, one more: **is `acceleration` correlated with `mpg`?** Careful — lower
acceleration numbers mean faster cars.